### ISSUE TITLE
Cursors are incredibly difficult to consume at times.

### ISSUE THREAD

**Issue by didinele** [porsager/postgres#185]
Was implementing an API route to retrieve data from a particular table via pagination.

As a general idea, this is incredibly difficult to do because everything needs to be done in the callback passed into `cursor`.

The only real way to do it right now is to create a promise, store its `resolve` method and make sure to block the callback until it resolves - resolving it on the next API request the user makes.

This ends up being incredibly painful, becomes hard to maintain and requires a lot of state. "Timing out cursors" (i.e no subsequent requests seem to be coming in so the transaction should be ended by returning `sql.END`) is also particularly difficult to do and the whole logic flow becomes very complicated and needy to not cause memory leaks.

I'd be interested in seeing this refactored (or, getting a new method).

The way I see it I should be able to:
- call `cursor` with a single parameter of `size`,
- be able to call `next` on the return value - which will output `size` rows as an array and a `done` boolean,
- extra points: implement an async iterator so this can be consumed as `for await (const rows of cursor(size))`

The proposed API would help significantly in my use case, as I would be able to store the result of `.cursor` and simply call `.next()` on every subsequent request without having to deal with crazy promises and state management between them.

**Comment by Minigugus** [comment]
Maybe you will be interested in a ponyfill AsyncIterator for sql``.cursor:

```ts
import postgres from 'postgres';

interface Deferable<T> {
  lock: Promise<T>;
  resolve: (value: T) => void;
  reject: (err: any) => void;
}

function deferable<T>(): Deferable<T> {
  let resolve: (value: T) => void, reject: (err: any) => void;
  return ({
    lock: new Promise<T>((res, rej) => {
      resolve = res;
      reject = rej;
    }),
    resolve: resolve!,
    reject: reject!
  })
}

function wrapCursor<T>(sql: postgres.Sql<any>, request: postgres.PendingQuery<T[]>, size: number) {
  let end = false;
  let cb2iter: Deferable<T[] | postgres.Sql<any>['END']> = deferable();
  let iter2cb: Deferable<void> = deferable();
  return Object.assign(
    request.cursor(size, async item => {
      await iter2cb.lock;
      iter2cb = deferable();
      cb2iter.resolve(item);
      cb2iter = deferable();
      return end;
    }),
    {
      [Symbol.asyncIterator]: () => ({
        async next() {
          if (end === false) {
            const { lock } = cb2iter;
            iter2cb.resolve();
            const value = await lock;
            if (Array.isArray(value)) // sql.END otherwise
              return ({ done: false, value });
          }
          return ({ done: true, value: undefined });
        },

        async return() {
          end = true;
          iter2cb.resolve();
          cb2iter.resolve(sql.END); // the `sql` parameter is only required because of this line
          return ({ done: true, value: undefined });
        },

        async throw(e?: any) {
          end = true;
          iter2cb.resolve();
          cb2iter.reject(e);
          return ({ done: true, value: undefined });
        }
      })
    } as AsyncIterable<T[]>
  )
};

// USAGE

const sql = postgres();

async function* paginate<T>(query: postgres.PendingQuery<T[]>, limit: number) {
  for await (const chunk of wrapCursor(sql, query, 100)) // caution: breaks with a chunk size of 1
    for (const item of chunk) {
      yield item;
      if (--limit)
        return; // automatically close the cursor
    }
}

for await (const item of paginate(sql<{ x: number }[]>`select * from generate_series(1, 15000) as x`, 1000)) {
  console.log(item);
}
```

This is a bit complex because of the locking mechanism but it might be a solution until it becomes available in `postgres`.

**Comment by porsager** [comment]
@didinele Would love to hear more about your specific use case, eg. are you sure you want to reserve a database connection for every single user for the duration of their paginated visits, and how would you delineate those?

**Comment by didinele** [comment]
Currently it's about the only solution I have found, but perhaps this may be due to my limited SQL/postgres knowledge.

In essence, I have a database table with potentially thousands of entries that are likely volatile, and each one needs to be represented in the UI.

My approach was taking an optional `cursor` query parameter into the route - if there is none, it means we're dealing with the first page, so I generate one and initiate the `cursor` call, ending the request with the rows returned and the newly generated cursor.

On subsequent requests I expect that `cursor` query parameter and use it to resolve the stored promise blocking the callback, sending the next chunk of rows.

This process is repeated until there is nothing left.

There's also a timeout running in the background from the first request that would cancel the transaction that is refreshed on every subsequent request.


In short, my use case is simply returning only 100 entries on every API call, no matter how I accomplish this. It is very well possible that I was wrong in attempting to use `cursor` to resolve this issue.

**Comment by porsager** [comment]
I see. 

There are different ways to do pagination depending on your requirements. I think this is a good overview worth reading - https://www.citusdata.com/blog/2016/03/30/five-ways-to-paginate/

Now using `sql.cursor` could be a valid solution, but you should be aware of the limitations and difficulties. 
The issue with using `sql.cursor` is that, as long as this cursor is active you are blocking a database connection. This means that, at best, only the same amount of users as there are database connections, will be abe to use the endpoint at the same time. This becomes really tricky because you then need to decide when a users cursor should be closed if the user is idle, so that the database connection can be made available to another user. Then if that user wasn't really idle after you closed the cursor, you need to either send the user back to the start (bad UI), or recreate a cursor with `offset` which was probably what you tried to prevent by using a cursor. This is the same hard problem as [cache invalidation](https://martinfowler.com/bliki/TwoHardThings.html), so if it's not a perfect match for your use case, I would avoid it.

The above reason is also why I tend to prefer the current API, because it matches the use cases i imagine cursors would be good for. I'd like to make it an async iterator as mentioned in the other issues at some point, but it's not high priority currently.

Hope this helps ;)

**Comment by didinele** [comment]
Alright, that makes a lot more sense now. Thanks a bunch!
